presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java

```diff
@@ -62,37 +63,46 @@ public PartitioningExchanger(
         }
         partitionGenerator = new LocalPartitionGenerator(hashGenerator, buffers.size());
 
-        partitionAssignments = new IntList[partitions.size()];
-        for (int i = 0; i < partitionAssignments.length; i++) {
-            partitionAssignments[i] = new IntArrayList();
-        }
+        pageBuilders = IntStream.range(0, buffers.size())
+                .mapToObj(i -> new PageBuilder(types))
+                .collect(toImmutableList());
     }
 
     @Override
     public synchronized void accept(Page page)
     {
-        // reset the assignment lists
-        for (IntList partitionAssignment : partitionAssignments) {
-            partitionAssignment.clear();
-        }
-
-        // assign each row to a partition
         for (int position = 0; position < page.getPositionCount(); position++) {
             int partition = partitionGenerator.getPartition(position, page);
-            partitionAssignments[partition].add(position);
+            PageBuilder pageBuilder = pageBuilders.get(partition);
+            appendRow(pageBuilder, page, position);
         }
 
-        // build a page for each partition
-        Block[] sourceBlocks = page.getBlocks();
-        Block[] outputBlocks = new Block[sourceBlocks.length];
-        for (int partition = 0; partition < buffers.size(); partition++) {
-            List<Integer> positions = partitionAssignments[partition];
-            if (!positions.isEmpty()) {
-                for (int i = 0; i < sourceBlocks.length; i++) {
-                    outputBlocks[i] = sourceBlocks[i].copyPositions(positions);
-                }
+        flush(false);
+    }
+
+    @Override
+    public void flush()
+    {
+        flush(true);
+    }
+
+    private void appendRow(PageBuilder pageBuilder, Page page, int position)
+    {
+        pageBuilder.declarePosition();
 
-                Page pageSplit = new Page(positions.size(), outputBlocks);
+        for (int channel = 0; channel < types.size(); channel++) {
+            Type type = types.get(channel);
+            type.appendTo(page.getBlock(channel), position, pageBuilder.getBlockBuilder(channel));
+        }
+    }
+
+    public void flush(boolean force)
+    {
+        for (int partition = 0; partition < pageBuilders.size(); partition++) {
+            PageBuilder partitionPageBuilder = pageBuilders.get(partition);
+            if (!partitionPageBuilder.isEmpty() && (force || partitionPageBuilder.isFull())) {
+                Page pageSplit = partitionPageBuilder.build();
+                partitionPageBuilder.reset();
                 memoryTracker.accept(pageSplit.getRetainedSizeInBytes());
                 buffers.get(partition).accept(new PageReference(pageSplit, 1, () -> memoryTracker.accept(-pageSplit.getRetainedSizeInBytes())));
             }
```

Comment:
You need to keep track of the memory that is in the non-flushed `PageBuilders`, too.